activerecord/lib/active_record/connection_adapters/abstract_adapter.rb

```diff
@@ -597,7 +602,7 @@ def reconnect!(restore_transactions: false)
         rescue => original_exception
           translated_exception = translate_exception_class(original_exception, nil, nil)
 
-          if retries_available > 0
+          if !query_timeout_reached?(start_time) && retries_available > 0
```

Comment:
I'd rather keep both "now" calls in the same method, at the cost of slight duplication of the subtraction / greater-than.